pkg/kubelet/images/image_manager_test.go

```diff
@@ -293,11 +296,140 @@ func TestPullAndListImageWithPodAnnotations(t *testing.T) {
+			imagePresent:      false,
+			pulledBySecret:    false,
+			ensuredBySecret:   false,
+			expectedWithFGOff: true,
+			expectedWithFGOn:  true,
+		},
+		{
+			description:       "PullAlways should always pull even if present and not pulled by secret",
+			container:         pullAlways,
+			imagePresent:      true,
+			pulledBySecret:    false,
+			ensuredBySecret:   false,
+			expectedWithFGOff: true,
+			expectedWithFGOn:  true,
+		},
+		{
+			description:       "PullAlways should always pull even if present and ensuredBySecret",
+			container:         pullAlways,
+			imagePresent:      true,
+			pulledBySecret:    true,
+			ensuredBySecret:   true,
+			expectedWithFGOff: true,
+			expectedWithFGOn:  true,
+		},
+		{
+			description:       "PullIfNotPresent should pull if not present",
+			container:         pullIfNotPresent,
+			imagePresent:      false,
+			pulledBySecret:    false,
+			ensuredBySecret:   false,
+			expectedWithFGOff: true,
+			expectedWithFGOn:  true,
+		},
+		{
+			description:       "PullIfNotPresent should pull if not present even if pulledBySecret and ensuredBySecret",
+			container:         pullIfNotPresent,
+			imagePresent:      false,
+			pulledBySecret:    true,
+			ensuredBySecret:   true,
+			expectedWithFGOff: true,
+			expectedWithFGOn:  true,
+		},
+		{
+			description:       "PullIfNotPresent should not pull if present (and secrets are not involved)",
+			container:         pullIfNotPresent,
+			imagePresent:      true,
+			pulledBySecret:    false,
+			ensuredBySecret:   false,
+			expectedWithFGOff: false,
+			expectedWithFGOn:  false,
+		},
+		{
+			description:       "PullIfNotPresent should not pull if present (and secrets are involved and match)",
+			container:         pullIfNotPresent,
+			imagePresent:      true,
+			pulledBySecret:    true,
+			ensuredBySecret:   true,
+			expectedWithFGOff: false,
+			expectedWithFGOn:  false,
+		},
+		{
+			description:       "PullIfNotPresent should pull if present and secrets are involved and no match, unless ensure fg is off",
+			container:         pullIfNotPresent,
+			imagePresent:      true,
+			pulledBySecret:    true,
+			ensuredBySecret:   false,
+			expectedWithFGOff: false,
+			expectedWithFGOn:  true,
+		},
+		{
+			description:       "PullNever should never report pull, but we'll throw error in EnsureImageExists()",
+			container:         pullNever,
+			imagePresent:      false,
+			pulledBySecret:    false,
+			ensuredBySecret:   false,
+			expectedWithFGOff: false,
+			expectedWithFGOn:  false,
+		},
+		{
+			description:       "PullNever should never pull even if pulled by secret and not ensured by secret",
+			container:         pullNever,
+			imagePresent:      true,
+			pulledBySecret:    true,
+			ensuredBySecret:   false,
+			expectedWithFGOff: false,
+			expectedWithFGOn:  false,
+		},
+	}
+
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KubeletEnsureSecretPulledImages, true)()
```

Comment:
Do we need the defer here? The true should be false in this case to disable the flag when the function exits. We don't need line 429 either to check to see if the feature gate is enabled, since it already should be.

Reply:
```
	t.Run("disabled_features", func(t *testing.T) {
		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KubeletEnsureSecretPulledImages, false)()
		for i, test := range tests {
			sp := shouldPullImage(test.container, test.imagePresent, test.pulledBySecret, test.ensuredBySecret)
			assert.Equal(t, test.expectedWithFGOff, sp, "TestCase[%d]: %s ensured image fg disabled", i, test.description)
		}
	})
	t.Run("enabled_features", func(t *testing.T) {
		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KubeletEnsureSecretPulledImages, true)()
		for i, test := range tests {
			sp := shouldPullImage(test.container, test.imagePresent, test.pulledBySecret, test.ensuredBySecret)
			assert.Equal(t, test.expectedWithFGOn, sp, "TestCase[%d]: %s ensured image fg enabled", i, test.description)
		}
	})
```
This case should test both enabled and disabled expected value.